File: sdks/go/docflow/extractor.go
package docflow

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/xgaslan/docflow/sdks/go/docflow/storage"
)

// Extractor handles PDF to Markdown extraction.
type Extractor struct {
	options Options
	storage storage.Storage
}

// NewExtractor creates a new Extractor instance.
func NewExtractor(opts ...ExtractorOption) *Extractor {
	e := &Extractor{
		options: DefaultOptions(),
	}

	for _, opt := range opts {
		opt(e)
	}

	return e
}

// ExtractorOption is a function that configures an Extractor.
type ExtractorOption func(*Extractor)

// WithExtractorOptions sets the extractor options.
func WithExtractorOptions(opts Options) ExtractorOption {
	return func(e *Extractor) {
		e.options = opts
	}
}

// WithExtractorStorage sets the storage backend for the extractor.
func WithExtractorStorage(s storage.Storage) ExtractorOption {
	return func(e *Extractor) {
		e.storage = s
	}
}

// ExtractToMarkdown extracts text from PDF and converts to Markdown.
func (e *Extractor) ExtractToMarkdown(ctx context.Context, pdfData []byte, filename string) (*ExtractResult, error) {
	if len(pdfData) == 0 {
		return nil, fmt.Errorf("PDF data is required")
	}

	// Ensure temp directory exists
	if err := os.MkdirAll(e.options.TempDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create temp directory: %w", err)
	}

	timestamp := time.Now().Unix()
	baseName := strings.TrimSuffix(filename, filepath.Ext(filename))
	safeName := sanitizeFilename(baseName)

	// Write PDF to temp file
	tempPDFPath := filepath.Join(e.options.TempDir, fmt.Sprintf("%s_%d.pdf", safeName, timestamp))
	if err := os.WriteFile(tempPDFPath, pdfData, 0644); err != nil {
		return nil, fmt.Errorf("failed to write temp PDF: %w", err)
	}
	defer os.Remove(tempPDFPath)

	// Extract text
	text, err := e.extractWithPdftotext(ctx, tempPDFPath)
	if err != nil {
		// Try alternative method
		text, err = e.extractBasic(ctx, tempPDFPath)
		if err != nil {
			return &ExtractResult{Success: false, Error: err}, nil
		}
	}

	// Get page count
	pageCount, _ := e.getPageCount(ctx, tempPDFPath)

	// Convert to markdown
	markdown := e.textToMarkdown(text, baseName)

	// Save to storage if configured
	var outputPath string
	if e.storage != nil {
		outputPath = fmt.Sprintf("%s_%d.md", safeName, timestamp)
		if err := e.storage.Save(outputPath, []byte(markdown)); err != nil {
			return nil, fmt.Errorf("failed to save markdown: %w", err)
		}
		outputPath = e.storage.GetURL(outputPath)
	}

	return &ExtractResult{
		Success:   true,
		Markdown:  markdown,
		FilePath:  outputPath,
		PageCount: pageCount,
	}, nil
}

// ExtractFromFile extracts markdown from a PDF file path.
func (e *Extractor) ExtractFromFile(ctx context.Context, path string) (*ExtractResult, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read PDF file: %w", err)
	}

	return e.ExtractToMarkdown(ctx, data, filepath.Base(path))
}

// GetPageCount returns the number of pages in a PDF.
func (e *Extractor) GetPageCount(ctx context.Context, pdfData []byte) (int, error) {
	// Write to temp file
	tempPath := filepath.Join(e.options.TempDir, fmt.Sprintf("pagecount_%d.pdf", time.Now().UnixNano()))
	if err := os.WriteFile(tempPath, pdfData, 0644); err != nil {
		return 0, err
	}
	defer os.Remove(tempPath)

	return e.getPageCount(ctx, tempPath)
}

// Preview extracts a preview of the first page.
func (e *Extractor) Preview(ctx context.Context, pdfData []byte, filename string) (*ExtractResult, error) {
	if len(pdfData) == 0 {
		return nil, fmt.Errorf("PDF data is required")
	}

	// Ensure temp directory exists
	if err := os.MkdirAll(e.options.TempDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create temp directory: %w", err)
	}

	timestamp := time.Now().Unix()
	baseName := strings.TrimSuffix(filename, filepath.Ext(filename))
	safeName := sanitizeFilename(baseName)

	// Write PDF to temp file
	tempPDFPath := filepath.Join(e.options.TempDir, fmt.Sprintf("%s_%d_preview.pdf", safeName, timestamp))
	if err := os.WriteFile(tempPDFPath, pdfData, 0644); err != nil {
		return nil, fmt.Errorf("failed to write temp PDF: %w", err)
	}
	defer os.Remove(tempPDFPath)

	// Get page count
	pageCount, _ := e.getPageCount(ctx, tempPDFPath)

	// Extract first page only
	text, err := e.extractFirstPage(ctx, tempPDFPath)
	if err != nil {
		text = "Preview not available"
	}

	markdown := e.textToMarkdown(text, baseName)

	// Truncate for preview
	if len(markdown) > 2000 {
		markdown = markdown[:2000] + "\n\n... (continued)"
	}

	return &ExtractResult{
		Success:   true,
		Markdown:  markdown,
		PageCount: pageCount,
	}, nil
}

func (e *Extractor) extractWithPdftotext(ctx context.Context, pdfPath string) (string, error) {
	ctx, cancel := context.WithTimeout(ctx, e.options.Timeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "pdftotext", "-layout", "-enc", "UTF-8", pdfPath, "-")
	output, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("pdftotext failed: %w", err)
	}

	return string(output), nil
}

func (e *Extractor) extractBasic(ctx context.Context, pdfPath string) (string, error) {
	ctx, cancel := context.WithTimeout(ctx, e.options.Timeout)
	defer cancel()

	// Try mutool
	cmd := exec.CommandContext(ctx, "mutool", "draw", "-F", "txt", pdfPath)
	output, err := cmd.Output()
	if err == nil {
		return string(output), nil
	}

	return "", fmt.Errorf("no PDF extraction tool available (install poppler-utils or mupdf-tools)")
}

func (e *Extractor) extractFirstPage(ctx context.Context, pdfPath string) (string, error) {
	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "pdftotext", "-f", "1", "-l", "1", "-layout", "-enc", "UTF-8", pdfPath, "-")
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}

	return string(output), nil
}

func (e *Extractor) getPageCount(ctx context.Context, pdfPath string) (int, error) {
	cmd := exec.CommandContext(ctx, "pdfinfo", pdfPath)
	output, err := cmd.Output()
	if err != nil {
		return 0, err
	}

	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		if strings.HasPrefix(line, "Pages:") {
			var pages int
			fmt.Sscanf(line, "Pages: %d", &pages)
			return pages, nil
		}
	}

	return 0, fmt.Errorf("could not determine page count")
}

func (e *Extractor) textToMarkdown(text, title string) string {
	var result strings.Builder

	// Add title
	result.WriteString("# ")
	result.WriteString(title)
	result.WriteString("\n\n")

	// Process text
	lines := strings.Split(text, "\n")
	var processedLines []string

	for _, line := range lines {
		line = strings.TrimSpace(line)

		if line == "" {
			if len(processedLines) > 0 && processedLines[len(processedLines)-1] != "" {
				processedLines = append(processedLines, "")
			}
			continue
		}

		// Detect headers
		if e.isPotentialHeader(line) {
			line = "## " + strings.Title(strings.ToLower(line))
		}

		// Detect bullet points
		if strings.HasPrefix(line, "•") || strings.HasPrefix(line, "●") || strings.HasPrefix(line, "○") {
			rest := strings.TrimPrefix(strings.TrimPrefix(strings.TrimPrefix(line, "•"), "●"), "○")
			line = "- " + strings.TrimSpace(rest)
		}

		// Detect numbered lists (already formatted, keep as is)
		if matched, _ := regexp.MatchString(`^\d+[\.]\s`, line); matched {
			// Keep numbered list format
		}

		processedLines = append(processedLines, line)
	}

	// Join with proper spacing
	inParagraph := false
	for i, line := range processedLines {
		if line == "" {
			if inParagraph {
				result.WriteString("\n\n")
				inParagraph = false
			}
			continue
		}

		if strings.HasPrefix(line, "##") || strings.HasPrefix(line, "- ") {
			if inParagraph {
				result.WriteString("\n\n")
			}
			result.WriteString(line)
			result.WriteString("\n")
			inParagraph = false
		} else {
			if inParagraph && i > 0 && processedLines[i-1] != "" {
				result.WriteString(" ")
			}
			result.WriteString(line)
			inParagraph = true
		}
	}

	result.WriteString("\n")
	return result.String()
}

func (e *Extractor) isPotentialHeader(line string) bool {
	if len(line) > 3 && len(line) < 60 && line == strings.ToUpper(line) {
		words := strings.Fields(line)
		if len(words) >= 2 {
			return true
		}
	}
	return false
}

File: sdks/python/docflow/extractor.py
"""PDF to Markdown extractor for DocFlow."""

import os
import re
import subprocess
import tempfile
import time
from pathlib import Path
from typing import Optional

from .storage.base import Storage
from .types import ExtractResult


class Extractor:
    """Extracts text from PDF files and converts to Markdown.
    
    Uses pdfminer.six for pure Python extraction, with fallback to pdftotext.
    
    Example:
        >>> extractor = Extractor()
        >>> with open("document.pdf", "rb") as f:
        ...     result = extractor.extract_to_markdown(f.read(), "document.pdf")
        >>> print(result.markdown)
    """

    def __init__(
        self,
        storage: Optional[Storage] = None,
        temp_dir: Optional[str] = None,
        use_native: bool = True,
    ) -> None:
        """Initialize the extractor.
        
        Args:
            storage: Storage backend for saving output files.
            temp_dir: Directory for temporary files.
            use_native: Use pdfminer.six (True) or pdftotext CLI (False).
        """
        self.storage = storage
        self.temp_dir = temp_dir or tempfile.gettempdir()
        self.use_native = use_native

    def extract_to_markdown(
        self, pdf_data: bytes, filename: str
    ) -> ExtractResult:
        """Extract text from PDF and convert to Markdown.
        
        Args:
            pdf_data: PDF file bytes.
            filename: Original filename.
            
        Returns:
            ExtractResult with markdown content.
        """
        if not pdf_data:
            return ExtractResult(success=False, error="PDF data is required")

        timestamp = int(time.time())
        base_name = Path(filename).stem
        safe_name = self._sanitize_filename(base_name)

        try:
            # Extract text
            if self.use_native:
                text = self._extract_with_pdfminer(pdf_data)
            else:
                text = self._extract_with_pdftotext(pdf_data, safe_name, timestamp)

            # Get page count
            page_count = self._get_page_count(pdf_data)

            # Convert to markdown
            markdown = self._text_to_markdown(text, base_name)

            # Save if storage configured
            output_path = None
            if self.storage:
                output_path = f"{safe_name}_{timestamp}.md"
                self.storage.save(output_path, markdown.encode("utf-8"))
                output_path = self.storage.get_url(output_path) or output_path

            return ExtractResult(
                success=True,
                markdown=markdown,
                file_path=output_path,
                page_count=page_count,
            )
        except Exception as e:
            return ExtractResult(success=False, error=str(e))

    def extract_from_file(self, path: str) -> ExtractResult:
        """Extract markdown from a PDF file path.
        
        Args:
            path: Path to PDF file.
            
        Returns:
            ExtractResult with markdown content.
        """
        pdf_data = Path(path).read_bytes()
        return self.extract_to_markdown(pdf_data, Path(path).name)

    def preview(self, pdf_data: bytes, filename: str) -> ExtractResult:
        """Get preview of extracted content (first page only).
        
        Args:
            pdf_data: PDF file bytes.
            filename: Original filename.
            
        Returns:
            ExtractResult with truncated markdown preview.
        """
        result = self.extract_to_markdown(pdf_data, filename)
        if result.success and len(result.markdown) > 2000:
            result.markdown = result.markdown[:2000] + "\n\n... (continued)"
        return result

    def get_page_count(self, pdf_data: bytes) -> int:
        """Get number of pages in PDF.
        
        Args:
            pdf_data: PDF file bytes.
            
        Returns:
            Page count.
        """
        return self._get_page_count(pdf_data)

    def _extract_with_pdfminer(self, pdf_data: bytes) -> str:
        """Extract text using pdfminer.six (pure Python)."""
        try:
            from pdfminer.high_level import extract_text
            from io import BytesIO
        except ImportError:
            raise ImportError("pdfminer.six is required: pip install pdfminer.six")

        return extract_text(BytesIO(pdf_data))

    def _extract_with_pdftotext(
        self, pdf_data: bytes, safe_name: str, timestamp: int
    ) -> str:
        """Extract text using pdftotext CLI."""
        temp_pdf = Path(self.temp_dir) / f"{safe_name}_{timestamp}.pdf"
        temp_pdf.write_bytes(pdf_data)

        try:
            result = subprocess.run(
                ["pdftotext", "-layout", "-enc", "UTF-8", str(temp_pdf), "-"],
                capture_output=True,
                text=True,
                timeout=60,
            )
            return result.stdout
        finally:
            temp_pdf.unlink(missing_ok=True)

    def _get_page_count(self, pdf_data: bytes) -> int:
        """Get page count from PDF."""
        try:
            from pdfminer.pdfpage import PDFPage
            from io import BytesIO
            
            return sum(1 for _ in PDFPage.get_pages(BytesIO(pdf_data)))
        except:
            return 0

    def _text_to_markdown(self, text: str, title: str) -> str:
        """Convert extracted text to Markdown format."""
        result = [f"# {title}\n\n"]

        lines = text.split("\n")
        processed_lines = []

        for line in lines:
            line = line.strip()

            if not line:
                if processed_lines and processed_lines[-1] != "":
                    processed_lines.append("")
                continue

            # Detect headers (ALL CAPS)
            if self._is_potential_header(line):
                line = "## " + line.title()

            # Detect bullet points
            for bullet in ["•", "●", "○"]:
                if line.startswith(bullet):
                    line = "- " + line[1:].strip()
                    break

            processed_lines.append(line)

        # Join with proper spacing
        in_paragraph = False
        for i, line in enumerate(processed_lines):
            if not line:
                if in_paragraph:
                    result.append("\n\n")
                    in_paragraph = False
                continue

            if line.startswith("##") or line.startswith("- "):
                if in_paragraph:
                    result.append("\n\n")
                result.append(line + "\n")
                in_paragraph = False
            else:
                if in_paragraph and i > 0 and processed_lines[i - 1]:
                    result.append(" ")
                result.append(line)
                in_paragraph = True

        result.append("\n")
        return "".join(result)

    def _is_potential_header(self, line: str) -> bool:
        """Check if line is likely a header."""
        return (
            3 < len(line) < 60
            and line == line.upper()
            and len(line.split()) >= 2
        )

    def _sanitize_filename(self, name: str) -> str:
        """Sanitize filename."""
        return re.sub(r'[/\\:*?"<>| ]', "_", name)
